### DISCUSSION TITLE
How to debug and fix

### DISCUSSION THREAD

**Discussion by SirAionTech** [sbidoul/ocadays2024-odoo-project-demo#3] [Q&A] [answered]
Hi,
thanks for the talk and for this PoC, I was wondering: how to manage contributions to OCA with this setup?

Let's say I have an `uv` installation and the customer finds a bug in a module: I re-create locally the customer's environment with `uv` and would like to debug/fix the code (obviously with the final goal to propose a PR in OCA).
The problem I see is that the code is deep within the virtual environment and there are no branches to work on.

**Comment by sbidoul** [comment, marked as answer]
@SirAionTech that is the part I had to skip in my talk.

The key is to do an editable install of the addon you want to debug.

So you git clone your fork of the OCA repo somewhere, you create a branch, then: `uv pip install -e /work/src/oca-repo/oca-addon`. (or `/work/src/oca-repo/setup/oca-addon` for Odoo <= 16).

You can then develop and debug as usual, then submit your PR.

Then, when the PR is up, you add the reference to the PR in `[tool.uv.sources]` like [this](https://github.com/sbidoul/ocadays2024-odoo-project-demo/compare/uv-level6...uv-level7), and `uv sync` to update the lock file. 

Of course, you can skip the editable install and play with `addons_path`, that will work too. I personally find the editable install easier and less error-prone.

**Reply by sbidoul** [reply]
Note that for the initial investigation, you can still debug what is in the venv site-packages. In vs code, you can add the site-packages/odoo directory to the workspace and set `justMyCode: false` in the debug config.

**Reply by SirAionTech** [reply]
Thanks, I installed
```
uv pip install -e /home/simone/work/odoo17/web/web_responsive
```
and it works great!

For the reference, it adds a new addons path at runtime (I only see it in the Odoo log)
```
/home/simone/work/odoo17/web/web_responsive/build/__editable__/odoo/addons
```
where the `build/__editable__/odoo/addons` directory is actually a symlink to `/home/simone/work/odoo17/web/web_responsive`.

**Reply by sbidoul** [reply]
> For the reference, it adds a new addons path at runtime

Yes, this is done by the [whool](https://github.com/sbidoul/whool) build backend.

**Reply by SirAionTech** [reply]
> > For the reference, it adds a new addons path at runtime
> 
> Yes, this is done by the [whool](https://github.com/sbidoul/whool) build backend.

I admit I don't know much about build backends, but I thought the build backend in this case was https://github.com/acsone/hatch-odoo:
https://github.com/sbidoul/ocadays2024-odoo-project-demo/blob/d7395acb4e8d3c4c387f4632aa0cf882150d933a/pyproject.toml#L5

How is https://github.com/sbidoul/whool involved?

**Reply by sbidoul** [reply]
Actually, `whool` is the build backend to package individual addons. You see it in the `pyproject.toml` next to the `__manifest__.py` in each OCA addon on branches 17.0 and 18.0.

`hatch-odoo` on the other hand, is an extension to the [hatchling](https://hatch.pypa.io/latest/config/build/#build-system) build backend, to package the end-customer project, as demonstrated in this demo repo..
